Translate `Components and Props` page (#45)
reactjs/pl.reactjs.org: content/docs/components-and-props.md

next: state-and-lifecycle.html
---

> Komponenty pozwalają ci podzielić interfejs użytkownika na niezależne, pozwalające na ponowne użycie części i myśleć o każdej z nich osobno. Ta strona wprowadza do pojęcia komponentów. Możesz znaleźć [szczegółowe odniesienie do API komponentów tutaj](/docs/react-component.html).

Reviewer: ```suggestion
Komponenty pozwalają podzielić interfejs użytkownika na niezależne, pozwalające na ponowne użycie części i myśleć o każdej z nich osobno. Ta strona wprowadza do pojęcia komponentów. W osobnym rozdziale opisaliśmy [szczegółową dokumentację API komponentów](/docs/react-component.html).
```

> Koncepcyjnie, komponenty są jak javascriptowe funkcje. Przyjmują one arbirtalne wartości wkładowe (nazywane "właściwościami") i zwracają reactowe elementy opisujące co powinno się pojawić na ekranie.

Reviewer: ```suggestion
Koncepcyjnie, komponenty są jak javascriptowe funkcje. Przyjmują one arbitralne wartości na wejściu (nazywane "właściwościami" (ang. *props*)) i zwracają reactowe elementy opisujące, co powinno się pojawić na ekranie.
```

## Komponenty funkcyjne i klasowe {#function-and-class-components}

Najprostszym sposobem na zdefiniowanie komponentu jest napisanie javascriptowej funkcji:

```js
function Welcome(props) {
  return <h1>Cześć, {props.name}</h1>;
}
```

> Ta funkcja jest uzasadnionym reactowym komponentem, bo przyjmuje pojedynczy argument "props" (który oznacza "właściwości" (ang. *properties*)) będący obiektem z danymi i zwraca reactowy element. Nazywamy takie komponenty "komponentami funkcyjnymi", bo są one dosłownie javascriptowymi funkcjami.

Reviewer: ```suggestion
Ta funkcja jest poprawnym reactowym komponentem, ponieważ przyjmuje pojedynczy argument "props" (który oznacza "właściwości" (z ang. *properties*)), będący obiektem z danymi, i zwraca reactowy element. Takie komponenty nazywamy "komponentami funkcyjnymi", gdyż są one dosłownie javascriptowymi funkcjami.
```

> Możesz również użyć [klasy ES6](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Classes) aby zdefiniować komponent:

Reviewer: ```suggestion
Do zdefiniowania komponentu możesz również użyć [klasy ze standardu ES6](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Classes):
```

```js
class Welcome extends React.Component {
  render() {
    return <h1>Cześć, {this.props.name}</h1>;
  }
}
```

> Dwa powyższe komponenty są równoważne z punktu widzenia Reacta.

Reviewer: ```suggestion
Obydwa powyższe komponenty są równoważne z punktu widzenia Reacta.
```

> Klasy mają kilka dodatkowych cech, które omówimy w [kolejnych rozdziałach](/docs/state-and-lifecycle.html). Na ten czas bedziemy używać komponentów funkcyjnych dla ich zwięzłości.

Reviewer: ```suggestion
Klasy mają kilka dodatkowych cech, które omówimy w [kolejnych rozdziałach](/docs/state-and-lifecycle.html). Na tego czasu będziemy używać komponentów funkcyjnych ze względu na ich zwięzły zapis.
```

Author: W drugim zdaniu chodzi o 'Na ten czas', czy 'Do tego czasu'?

Reviewer: Chodziło mi o "Do tego czasu" :D

Author: 36a975f3b4d5c69ab926942f930ed0862d228b2b

## Renderowanie komponentu {#rendering-a-component}

> Poprzednio, napotykaliśmy reactowe elementy, które reprezentowały tagi drzewa DOM:

Reviewer: ```suggestion
Poprzednio napotykaliśmy reactowe elementy, które reprezentowały znaczniki DOM:
```

```js
const element = <div />;
```

> Natomiast elementy mogą również reprezentować komponenty zdefiniowane przez użytkownika:

Reviewer: ```suggestion
Elementy mogą również reprezentować komponenty zdefiniowane przez użytkownika:
```

```js
const element = <Welcome name="Sara" />;
```

> Kiedy React widzi element reprezentujący komponent zdefiniowany przez użytkownika, przekazuje on do niego JSX'owe atrybuty jako jeden obiekt. Nazywamy ten obiekt "właściwościami".

Reviewer: ```suggestion
Kiedy React widzi element reprezentujący komponent zdefiniowany przez użytkownika, przekazuje do niego JSX-owe atrybuty jako jeden obiekt. Obiekt ten nazywamy "właściwościami" komponentu.
```

> Na przykład, ten kod renderuje "Cześć, Sara" na stronie:

Reviewer: ```suggestion
Dla przykładu, poniższy kod renderuje na stronie napis "Cześć, Sara":
```

[](codepen://components-and-props/rendering-a-component)

Podsumujmy, co dzieje się w tym przykładzie:

1. Wywołujemy `ReactDOM.render()` z elementem `<Welcome name="Sara" />`.

> 2. React wywołuje komponent `Welcome` z `{name: 'Sara'}` jako właściwości.

Reviewer: ```suggestion
2. React wywołuje komponent `Welcome` z właściwościami `{name: 'Sara'}`.
```

> 3. Nasz komponent `Welcome` zwraca jako wynik element `<h1>Hello, Sara</h1>`.

Reviewer: ```suggestion
3. Nasz komponent `Welcome` jako wynik zwraca element `<h1>Cześć, Sara</h1>`.
```

> 4. React DOM wydajnie uaktualinia drzewo DOM, aby równało się `<h1>Hello, Sara</h1>`.

Reviewer: ```suggestion
4. React DOM w optymalny sposób aktualizuje drzewo DOM, aby odpowiadało elementowi `<h1>Cześć, Sara</h1>`.
```

> >**Wskazówka:** Zawsze zaczynaj nazwy komponentów od dużej litery.

Reviewer 2: To ewentualnie zmieniłbym na `Uwaga`, tak jak np. tu: https://github.com/reactjs/pl.reactjs.org/pull/40/files#diff-6024d880e5eef39cfbff54653ef34fb6R61
`Wskazówka` nie jest właściwie niepoprawna, ale bardziej pasuje do angielskiego `Tip`.

```suggestion
>**Uwaga:** Zawsze zaczynaj nazwy komponentów od dużej litery.
```

>

> >React traktuje komponenty zaczynające się od małej litery jako tagi drzewa DOM. Na przykład, `<div />` reprezentuje HTML'owy tag div, ale `<Welcome />` reprezentuje komponent i wymaga, aby `Welcome` było w zasięgu.

Reviewer: ```suggestion
>React traktuje komponenty zaczynające się od małej litery jako tagi drzewa DOM. Na przykład, `<div />` reprezentuje HTML-owy znacznik 'div', ale już `<Welcome />` reprezentuje komponent i wymaga, aby `Welcome` było w zasięgu (ang. *scope*).
```

>

> >Aby dowiedzieć się więcej o uzasadnieniu tej konwencji, przeczytaj [dogłębną analizę JSX](/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized).

Reviewer: ```suggestion
>Aby dowiedzieć się więcej o uzasadnieniu tej konwencji, przeczytaj [dogłębną analizę składni JSX](/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized).
```

## Kompozycja komponentów {#composing-components}

> Komponenty mogą odwoływać się do innych komponentów i ich wyników. To pozwala nam na używanie tej samej abstrakcji komponentu na dowolnym poziomie szczegółowości.

Reviewer: ```suggestion
Komponenty przy zwracaniu wyniku mogą mogą odwoływać się do innych komponentów. Pozwala to używać tej samej abstrakcji komponentu na dowolnym poziomie szczegółowości. Przycisk, formularz, okno dialogowe, ekran - w aplikacjach reactowych tego typu składniki są zwykle reprezentowane przez dedykowane komponenty.
```

> Na przykład możemy stworzyć komponent `App`, który renderuje `Welcome` wiele razy:

Reviewer: ```suggestion
Możemy dla przykładu stworzyć komponent `App`, który wielokrotnie renderuje komponent `Welcome`:
```

```js{8-10}
function Welcome(props) {
  return <h1>Cześć, {props.name}</h1>;
}

> function App() {

Reviewer: Proponuję przystosować imiona do polskich realiów. Sara, Karol, Edyta? :)

Author: Z tego co widzę wszędzie jest Sara, ale może coś przeoczyłem

[](codepen://components-and-props/composing-components)

> Zazwyczaj, nowe reactowe aplikacje mają pojedynczy komponent `App` u samej góry. Natomiast, jeśli będziesz integrował Reacta z istniejącą aplikacją, możesz zacząć od dołu z małymi komponentami jak na przykład `Button` i stopniowo wypracować sobie drogę na szczyt hierarchii widoku.

Reviewer: ```suggestion
Nowe aplikacje reactowe na samej górze drzewa zazwyczaj renderują pojedynczy komponent `App`. Jeśli jednak musisz zintegrować Reacta z istniejącą aplikacją, możesz zacząć od samego dołu, dodając niewielkie komponenty (np. `Button`) i stopniowo przepisywać całą strukturę aż do samej góry.
```

## Wyciąganie komponentów {#extracting-components}

> Nie bój się dzielenia komponentów na mniejsce komponenty.

Reviewer: ```suggestion
Nie bój się dzielenia komponentów na mniejsze części.
```

> Na przykład rozważ ten komponent `Commment`:

Reviewer: ```suggestion
Rozważ poniższy komponent `Commment`:
```

[](codepen://components-and-props/extracting-components)

> Przyjmuje on `autora` (ang. *author*) - obiekt, `text` - string i `datę` (ang. *date*) jako właściwości, oraz opisuje komentarz na stronie mediów społecznościowych.

Reviewer: ```suggestion
Przyjmuje on obiekt `author`, napis `text` i datę `date` jako właściwości i zwraca strukturę opisującą komentarz na portalu mediów społecznościowych.
```

> Zmiana tego komponentu może być zawiła z powodu całego tego zagnieżdżenia, a ponowne użycie jego indywidulnych części jest trudne. Wyciągnijmy z niego kilka komponentów.

Reviewer: ```suggestion
Zmiana tego komponentu czy ponowne użycie jego poszczególnych części może okazać się skomplikowane z powodu całego tego zagnieżdżenia. Rozbijmy go zatem na kilka mniejszych komponentów.
```

> Najpierw, wyciągniemy `Avatar`:

Reviewer: ```suggestion
Najpierw wydzielmy komponent `Avatar`:
```

}
```

> `Avatar` nie musi wiedzieć, że jest renderowany w środku komponentu `Comment`. Dlatego też daliśmy jego właściwości bardziej ogólną nazwę: `użytkownik` (ang. *user*) zamiast `autor`.

Reviewer: ```suggestion
`Avatar` nie musi wiedzieć, że jest renderowany wewnątrz komponentu `Comment`. Dlatego też daliśmy jego właściwości bardziej ogólną nazwę `user` zamiast `author`.
```

> Polecamy nazywać właściwości z punktu widzenia komponentu, a nie kontekstu, w którym jest używany.

Reviewer: ```suggestion
Zalecamy nadawanie nazw właściwościom z punktu widzenia komponentu, a nie kontekstu, w którym jest używany.
```

> Możemy teraz dodać do komponentu `Comment` drobne uproszczenie:

Reviewer: ```suggestion
Możemy teraz uprościć nieco komponent `Comment`:
```

}
```

> Następnie wyciągniemy komponent `UserInfo`, który renderuje `Avatar` obok nazwy użytkownika:

Reviewer: ```suggestion
Następnie wydzielmy komponent `UserInfo`, który wyrenderuje `Avatar` obok nazwy użytkownika:
```

[](codepen://components-and-props/extracting-components-continued)

> Wyciąganie komponentów może z początku wydawać się żmudnym zajęciem, ale posiadanie palety pozwalających na ponowne użycie komponentów jest opłacalne w większych aplikacjach. Dobrą praktyczną zasadą jest, że jeśli część twojego interfejsu użytkownika jest używana wiele razy (`Button`, `Panel`, `Avatar`) lub jest jest ona dostatecznie skomplikowana sama w sobie (`App`, `FeedStory`, `Comment`), jest ona dobrym kandydatem do stania się komponentem wielokrotnego użytku.

Reviewer: ```suggestion
Wyciąganie komponentów może z początku wydawać się żmudnym zajęciem, ale posiadanie palety pozwalających na ponowne użycie komponentów jest opłacalne w większych aplikacjach. Dobrą praktyczną zasadą jest, że jeśli część twojego interfejsu użytkownika jest używana wielokrotnie (np. `Button`, `Panel`, `Avatar`) lub jest ona dostatecznie skomplikowana sama w sobie (np. `App`, `FeedStory`, `Comment`), jest ona dobrym kandydatem do stania się komponentem wielokrotnego użytku.
```

## Właściwości są tylko do odczytu {#props-are-read-only}

> Bez względu na to, czy zadeklarujesz komponent [jako funkcja czy klasa](#function-and-class-components), nie może on nigdy modyfikować swoich właściwości. Rozważ tą funkcję `sum`:

Reviewer: ```suggestion
Bez względu na to, czy zadeklarujesz komponent [jako funkcję czy klasę](#function-and-class-components), nie może on nigdy modyfikować swoich właściwości. Rozważ następującą funkcję `sum`:
```

```js
function sum(a, b) {
  return a + b;
}
```

> Funkcje tego typu nazywane są ["czystymi"](https://en.wikipedia.org/wiki/Pure_function), dlatego że nie próbują one zmieniać swoich wartości wejściowych i zawsze zwracają ten sam wynik dla tych samych wartości wejściowych.

Reviewer: ```suggestion
Funkcje tego typu nazywane są ["czystymi"](https://en.wikipedia.org/wiki/Pure_function) (ang. *pure function*), dlatego że nie próbują one zmieniać swoich argumentów i zawsze zwracają ten sam wynik dla tych samych argumentów.
```

> W przeciwieństwie do poprzedniej, ta funkcja nie jest czysta, bo zmienia swoją wartość wejściową.

Reviewer: ```suggestion
W przeciwieństwie do poprzedniej funkcji, ta poniżej nie jest "czysta", ponieważ zmienia swój argument.
```

```js
function withdraw(account, amount) {
  account.total -= amount;
}
```

React jest bardzo elastyczny, ale ma jedną ścisłą zasadę:

**Wszytkie komponenty muszą zachowywać się jak czyste funkcje w odniesieniu do ich właściwości.**

> Oczywiście, interfejsy użytkownika aplikacji są dynamiczne i zmieniają się w czasie. W [kolejnym rozdziale](/docs/state-and-lifecycle.html), wprowadzimy nowe pojęcie "stanu". Stan pozwala reactowym komponentom na zmianę swojego wyniku w czasie w odpowiedzi na akcje użytkownika, żądania sieciowe i wszystko inne, bez naruszania tej zasady.

Reviewer: ```suggestion
Rzecz jasna, interfejsy użytkownika w aplikacjach są zwykle dynamiczne, zmieniają się w czasie. W [kolejnym rozdziale](/docs/state-and-lifecycle.html), wprowadzimy nowe pojęcie "stanu". Stan pozwala reactowym komponentom na zmianę swojego wyniku w czasie, w odpowiedzi na akcje użytkownika, żądania sieciowe itp. bez naruszania powyższej zasady.
```

Author: 36a975f3b4d5c69ab926942f930ed0862d228b2b

> Rzecz jasna, interfejsy użytkownika w aplikacjach są zwykle dynamiczne, zmieniają się w czasie. W [kolejnym rozdziale](/docs/state-and-lifecycle.html) wprowadzimy nowe pojęcie "stanu". Stan pozwala reactowym komponentom na zmianę swojego wyniku w czasie, w odpowiedzi na akcje użytkownika, żądania sieciowe itp. bez naruszania powyższej zasady.

Reviewer 2: ```suggestion
Rzecz jasna, interfejsy użytkownika w aplikacjach są zwykle dynamiczne, zmieniają się w czasie. W [kolejnym rozdziale](/docs/state-and-lifecycle.html) wprowadzimy nowe pojęcie "stanu". Stan pozwala komponentom reactowym na zmianę swojego wyniku w czasie, w odpowiedzi na akcje użytkownika, żądania sieciowe itp. bez naruszania powyższej zasady.
```
